src/Raven.Server/Utils/MetricCacher.cs

```diff
@@ -67,86 +68,55 @@ public T GetValue<T>(string key, Func<T> factory = null)
         {
-            private readonly Func<T> _factory;
+            private readonly TimeSpan _refreshRate;
+            private Func<object> _factory;
+            private Task<DateTime> _task;
+            private object _value;
 
-            public MetricValue(TimeSpan refreshRate, Func<T> factory)
-                : base(refreshRate)
+            public MetricValue(TimeSpan refreshRate, Func<object> factory)
             {
-                if (factory == null)
-                    throw new ArgumentNullException(nameof(factory));
-
+                _refreshRate = refreshRate;
                 _factory = factory;
+                _value = factory();
+                _task = Task.FromResult(SystemTime.UtcNow + _refreshRate);
             }
 
-            protected override object RefreshInternal()
-            {
-                return _factory();
-            }
-        }
 
-        private abstract class MetricValueBase
-        {
-            private DateTime _lastRefresh;
-            private TimeSpan _refreshRate;
-
-            protected MetricValueBase(TimeSpan refreshRate)
+            public object GetRefreshedValue()
             {
-                _refreshRate = refreshRate;
-            }
+                var currentTask = _task;
+                if (currentTask.IsCompleted == false)
+                    return _value; // return current value, while it is being computed
 
-            public object Value { get; private set; }
+                if (currentTask.IsFaulted)
+                {
+                    throw new InvalidOperationException("Failed to get value", currentTask.Exception);
+                }
 
-            protected abstract object RefreshInternal();
+                var nextRefreshForTask = currentTask.Result;
+                if (SystemTime.UtcNow < nextRefreshForTask)
+                    return _value;// no need to refresh yet...
 
-            internal void Refresh()
-            {
-                try
+                var task = new Task<DateTime>(() =>
                 {
-                    Value = RefreshInternal();
-                }
-                finally
+                    var result = _factory();
+                    var nextRefresh = SystemTime.UtcNow + _refreshRate;
+                    Interlocked.Exchange(ref _value, result);
+                    return nextRefresh;
+                });
+                // try to update the task
+                if (Interlocked.CompareExchange(ref _task, task, currentTask) == currentTask)
                 {
-                    _lastRefresh = SystemTime.UtcNow;
+                    // we won the right to start the task
+                    task.Start();
                 }
-            }
-
-            internal bool ShouldRefresh(out bool first)
-            {
-                first = _lastRefresh == DateTime.MinValue;
 
-                return first || SystemTime.UtcNow - _lastRefresh >= _refreshRate;
+                return _value;
```

Comment:
You might return the stale value here because the just started task could not update it yet. I'm not sure if that matters much but this is a change in comparison to the current behavior.

Reply:
Yes, that is by design, because we are currently still processing it.
We are okay in returning a stale value, I believe.